src/lib.rs

```diff
@@ -622,6 +668,14 @@ pub enum Error {
     InvalidPadding,
     /// The whole string must be of one case
     MixedCase,
+    /// Writing UTF-8 data failed
+    WriteFailure(fmt::Error),
```

Comment:
I don't think this is the right way of using `fmt::Error`. It's just a marker that something failed and you're supposed to retrieve the error fro underlying stream. In case of `std::io` it's done for you and in case of `String` there's no error so unwrapping is the correct thing to do.

This library abuses the `fmt::Write` API to signal display failure but `Display` must not fail for any other reason than formatter failing.

From [the docs](https://doc.rust-lang.org/std/fmt/index.html):

> Formatting implementations should ensure that they propagate errors from the [Formatter](https://doc.rust-lang.org/std/fmt/struct.Formatter.html) (e.g., when calling [write!](https://doc.rust-lang.org/std/macro.write.html)). However, they should never return errors spuriously. That is, a formatting implementation must and may only return an error if the passed-in [Formatter](https://doc.rust-lang.org/std/fmt/struct.Formatter.html) returns an error. This is because, contrary to what the function signature might suggest, string formatting is an infallible operation. This function only returns a result because writing to the underlying stream might fail and it must provide a way to propagate the fact that an error has occurred back up the stack.

Reply:
Good find! This will definitely inform our API overhaul.